src/UserContext/AnonymousRequestMatcher.php

```diff
@@ -13,36 +13,62 @@
 
 use Symfony\Component\HttpFoundation\Request;
 use Symfony\Component\HttpFoundation\RequestMatcherInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
 
 /**
  * Matches anonymous requests using a list of identification headers.
  */
 class AnonymousRequestMatcher implements RequestMatcherInterface
 {
-    private $sessionNamePrefix;
+    /**
+     * The options configured in the constructor argument or default values.
+     *
+     * @var array
+     */
+    protected $options;
 
     /**
-     * @param string $sessionNamePrefix Prefix for session cookies. Must match your PHP session configuration
+     * When creating this matcher, you can configure a number of options.
+     *
+     * - user_identifier_headers: List of request headers that authenticate a non-anonymous request.
+     * - session_name_prefix:     Prefix for session cookies. Must match your PHP session configuration.
+     *
+     * @param array $options Options to overwrite the default options
+     *
+     * @throws \InvalidArgumentException if unknown keys are found in $options
      */
-    public function __construct($sessionNamePrefix)
+    public function __construct(array $options = array())
     {
-        $this->sessionNamePrefix = $sessionNamePrefix;
+        $resolver = new OptionsResolver();
+        $resolver->setDefaults(array(
+            'user_identifier_headers' => array('Cookie', 'Authorization', 'HTTP_AUTHORIZATION', 'PHP_AUTH_USER'),
+            'session_name_prefix' => 'PHPSESSID',
+        ));
+
+        $this->options = $resolver->resolve($options);
     }
 
     public function matches(Request $request)
     {
         // You might have to enable rewriting of the Authorization header in your server config or .htaccess:
         // RewriteEngine On
         // RewriteRule .* - [E=HTTP_AUTHORIZATION:%{HTTP:Authorization}]
-        if ($request->server->has('AUTHORIZATION') ||
-            $request->server->has('HTTP_AUTHORIZATION') ||
-            $request->server->has('PHP_AUTH_USER')
-        ) {
-            return false;
+        foreach ($this->options['user_identifier_headers'] as $header) {
+            if ($request->headers->has($header)) {
+                if ('cookie' === strtolower($header) &&
+                    (0 === $request->cookies->count() || !empty($this->options['session_name_prefix']))
+                ) {
+                    // ignore empty cookie header
+                    // ignore cookie header when session_name_prefix is set
+                    continue;
+                }
+
+                return false;
+            }
         }
 
         foreach ($request->cookies as $name => $value) {
```

Comment:
the cookies header is hardcoded here - which i think is correct. makes me think we should not pass around Cookies in the list of authorization headers at all.  and then we can get rid of the checks in the loop above.